include/irr/asset/IAsset.h

```diff
@@ -6,155 +6,166 @@
+
+		public:
+			//! this could actually be reworked to something more usable
+			virtual const char* getLoaderName() = 0;
+		};
+
+		class IAsset : virtual public core::IReferenceCounted
+		{
+		public:
+			enum E_TYPE : uint64_t
+			{
+				//! asset::ICPUBuffer
+				ET_BUFFER = 1u << 0u,
+				//! asset::CImageData - maybe rename to asset::CSubImageData
+				ET_SUB_IMAGE = 1u << 1u,
+				//! asset::ICPUTexture
+				ET_IMAGE = 1u << 2u,
+				//! asset::ICPUMeshBuffer
+				ET_SUB_MESH = 1u << 3u,
+				//! asset::ICPUMesh
+				ET_MESH = 1u << 4u,
+				//! asset::ICPUSkeleton - to be done by splitting CFinalBoneHierarchy
+				ET_SKELETON = 1u << 5u,
+				//! asset::ICPUKeyframeAnimation - from CFinalBoneHierarchy
+				ET_KEYFRAME_ANIMATION = 1u << 6u,
+				//! asset::ICPUShader
+				ET_SHADER = 1u << 7u,
+				//! asset::ICPUSpecializedShader
+				ET_SPECIALIZED_SHADER = 1u << 8,
+				//! asset::ICPUMeshDataFormatDesc
+				ET_MESH_DATA_DESCRIPTOR = 1u << 8,
+				//! reserved, to implement later
+				ET_GRAPHICS_PIPELINE = 1u << 9u,
+				//! reserved, to implement later
+				ET_SCENE = 1u << 10u,
+				//! lights, etc.
+				ET_IMPLEMENTATION_SPECIFIC_METADATA = 1u << 31u
+				//! Reserved special value used for things like terminating lists of this enum
+			};
+			constexpr static size_t ET_STANDARD_TYPES_COUNT = 12u;
+
+			static uint32_t typeFlagToIndex(E_TYPE _type)
+			{
+				uint32_t type = (uint32_t)_type;
+				uint32_t r = 0u;
+				while (type >>= 1u) ++r;
+				return r;
+			}
+
+			IAsset() : isDummyObjectForCacheAliasing{ false }, m_metadata{ nullptr } {}
+
+			virtual size_t conservativeSizeEstimate() const = 0;
+
+			IAssetMetadata* getMetadata() { return m_metadata.get(); }
+			const IAssetMetadata* getMetadata() const { return m_metadata.get(); }
+
+			friend IAssetManager;
+
+		private:
+			core::smart_refctd_ptr<IAssetMetadata> m_metadata;
+
+			void setMetadata(IAssetMetadata* _metadata) { m_metadata = _metadata; }
+
+		protected:
+			bool isDummyObjectForCacheAliasing;
+			//! To be implemented by base classes, dummies must retain references to other assets
+			//! but cleans up all other resources which are not assets.
+			virtual void convertToDummyObject() = 0;
+			//! Checks if the object is either not dummy or dummy but in some cache for a purpose
+			inline bool isInValidState() { return !isDummyObjectForCacheAliasing /* || !isCached TODO*/; }
+			//! Pure virtual destructor to ensure no instantiation
+			virtual ~IAsset() = 0;
+
+		public:
+			//! To be implemented by derived classes
+			virtual E_TYPE getAssetType() const = 0;
+			//! 
+			inline bool isADummyObjectForCache() const { return isDummyObjectForCacheAliasing; }
+		};
+
+		class SAssetBundle
+		{
+			using contents_container_t = core::refctd_dynamic_array<core::smart_refctd_ptr<IAsset>>;
+		public:
+			SAssetBundle(std::initializer_list<core::smart_refctd_ptr<IAsset>> _contents = {}) : m_contents(new contents_container_t(_contents), core::dont_grab)
```

Comment:
@Crisspl and @Przemog1 you know that there's a `core::make_smart_refctd_ptr` so you don't need to use the nasty `dont_grab` constructor